### ISSUE TITLE
Casting list of JSONObject to JSONObject throws ClassCastException

### ISSUE THREAD

**Issue by ggavriilidis** [stleary/JSON-java#491]
When calling the toList() method of a JSONArray which contains JSONObject the returned list contains HashMap elements.
The documentation mentions:

> Returns a java.util.List containing all of the elements in this array.
> If an element in the array is a JSONArray or JSONObject it will also be converted

However, the elements are not converted to List<JSONObject>.
Casting to JSONObject throws an exception if the toList() method is used.
Below is an example of the problem occurring:

### Steps to reproduce

Run the below unit test:

 ```
@Test
    public void shouldCastArrayOfJSONObjectsToList() {
        String a = "{\"a\": 1}";
        String b = "{\"b\":2}";
        String jsonString = "{\"list\":[" + a + "," + b + "]}";
        JSONObject jsonObject = new JSONObject(jsonString);

        JSONArray jsonArray = jsonObject.getJSONArray("list");

        List<JSONObject> jsonObjects = jsonArray.toList().stream().map(JSONObject.class::cast).collect(Collectors.toList());

        assertEquals(jsonObjects.size(), 2);
    }
```

### Expected Behaviour

No exception should be thrown and the test should pass

### Actual Behaviour

ClassCastException is thrown.

```
java.lang.ClassCastException: Cannot cast java.util.HashMap to org.json.JSONObject

	at java.lang.Class.cast(Class.java:3369)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at JsonTest.shouldCastArrayOfJSONObjectsToList(JsonTest.java:21)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

### Using iterator works

The following test however passes:

```
@Test
    public void shouldCastArrayOfJSONObjectsToList() {
        String a = "{\"a\": 1}";
        String b = "{\"b\":2}";
        String jsonString = "{\"list\":[" + a + "," + b + "]}";
        JSONObject jsonObject = new JSONObject(jsonString);

        JSONArray jsonArray = jsonObject.getJSONArray("list");

        List<JSONObject> jsonObjects = Lists.newArrayList(jsonArray.iterator()).stream().map(JSONObject.class::cast).collect(Collectors.toList());

        assertEquals(jsonObjects.size(), 2);
    }
```

Environment Information:
Operating System: macOS Sierra v 10.12.6
JDK Version: 1.8.0_152
org.json version: 20190722

**Comment by johnjaylward** [comment]
The `toList` method does not convert to a list of `JSONObject`. It converts both `JSONArray` and `JSONObject` to standard java types `List` and `Map`. For a general use case, you would use something like this:

```java
public void shouldCastArrayOfJSONObjectsToList() {
       String a = "{\"a\": 1}";
       String b = "{\"b\":2}";
       String jsonString = "{\"list\":[" + a + "," + b + ",8" + "]}";
       JSONObject jsonObject = new JSONObject(jsonString);

       JSONArray jsonArray = jsonObject.getJSONArray("list");

       List<Object> objects = jsonArray.toList();

       assertEquals(objects.size(), 3); // list of Map, Map, Integer
    }
```

For your specific example, if your `List` would not be of JSONObject, but instead `Map` objects:
```java
public void shouldCastArrayOfJSONObjectsToList() {
       String a = "{\"a\": 1}";
       String b = "{\"b\":2}";
       String jsonString = "{\"list\":[" + a + "," + b + "]}";
       JSONObject jsonObject = new JSONObject(jsonString);

       JSONArray jsonArray = jsonObject.getJSONArray("list");

       List<Map<String,Object>> objects = jsonArray.toList().stream().map(Map.class::cast).collect(Collectors.toList());

       assertEquals(objects.size(), 2); // list of Map, Map
    }
```

However, If you really want `JSONObject`s, I would use this method:
```java
public void shouldCastArrayOfJSONObjectsToList() {
       String a = "{\"a\": 1}";
       String b = "{\"b\":2}";
       String jsonString = "{\"list\":[" + a + "," + b + "]}";
       JSONObject jsonObject = new JSONObject(jsonString);

       JSONArray jsonArray = jsonObject.getJSONArray("list");

       List<Map<String,Object>> objects = StreamSupport.stream(jsonArray.spliterator(), false).map(o -> {
            if(o instanceof JSONObject) {
                return (JSONObject)o;
            }
            return null;
        }).filter(Objects::nonNull)
        .collect(Collectors.toList());;

       assertEquals(objects.size(), 2); // list of Map, Map
    }
```

The last method prevents a lot of work that the `toList` method is performing and will also filter out non-Json objects from the list.

**Comment by ggavriilidis** [comment]
Thanks for your response @johnjaylward . 

Is it expected that the JSONArray iterator and spliterator methods behave differently than the toList method?

If this is expected maybe the documentation could add some more clarification.
Currently it mentions:

```
Returns a java.util.List containing all of the elements in this array.
If an element in the array is a JSONArray or JSONObject it will also be converted.
```

Could it be saying instead: 

```
Returns a java.util.List containing all of the elements in this array.
If an element in the array is a JSONArray or JSONObject it will also be converted to a List and a Map respectively.
```

This clarifies what the output of the method is.

**Comment by stleary** [comment]
Fixed in #492
